<a href="https://colab.research.google.com/github/FlyAIBox/Agent_In_Action/blob/main/05-agent-model-finetuning/lora/lora-from-peft-in-google-colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🚀 使用 Hugging Face PEFT 库实现 LoRA 微调

## 📖 教程概述

本教程将展示如何使用 **Hugging Face PEFT（Parameter-Efficient Fine-Tuning）** 库实现 LoRA 微调，这是工业界推荐的标准方法。

### 🎯 为什么使用 PEFT 库？

#### ✅ PEFT 库的优势

1. **工业级实现**：经过大规模验证，稳定可靠
2. **易于使用**：几行代码即可应用 LoRA
3. **功能丰富**：支持 LoRA、LoHa、LoKr 等多种方法
4. **生态集成**：与 Transformers、Accelerate 无缝集成
5. **持续维护**：Hugging Face 官方支持

#### 📊 与手动实现的对比

| 特性 | 手动实现 | PEFT 库 |
|:-----|:-------:|:-------:|
| **代码量** | 多（需实现 LoRA 层） | 少（几行配置） |
| **学习价值** | 深入理解原理 | 快速应用实践 |
| **维护成本** | 高（需自己维护） | 低（官方维护） |
| **功能完整性** | 基础功能 | 丰富功能 |
| **错误风险** | 较高 | 较低 |
| **生产可用性** | 需额外验证 | 工业级稳定 |
| **适用场景** | 学习研究 | 生产环境 |

### 🎓 学习路径建议

1. **第一步**：学习 PEFT 库（掌握工具）
2. **第二步**：学习手动实现（理解原理）
3. **第三步**：实际项目中使用 PEFT

---

## 💻 环境要求

### 🌐 Google Colab 环境（推荐新手）

本教程已在 Google Colab 环境测试通过：

| 配置项 | Google Colab 规格 |
|:------|:-----------------|
| **平台** | Google Colab (免费版/Pro) |
| **操作系统** | Ubuntu 22.04 LTS |
| **Python 版本** | 3.12 |
| **GPU 型号** | Tesla T4 |
| **GPU 显存** | 16 GB |
| **CUDA 版本** | 12.6 |
| **PyTorch 版本** | 2.9.0 |

💡 **Google Colab 优势**：
- ✅ 免费提供 GPU 资源（T4 16GB）
- ✅ 预装深度学习环境，开箱即用
- ✅ 无需本地配置，浏览器即可运行
- ⚠️ 会话有时间限制（12小时），需定期保存

### 🏠 私有环境部署要求

如果您需要在本地或私有云环境运行本教程，建议满足以下配置：

#### 最低配置要求

| 配置项 | 最低要求 | 推荐配置 |
|:------|:--------|:---------|
| **操作系统** | Ubuntu 20.04+ / CentOS 8+ / Windows 10+ | Ubuntu 22.04 LTS |
| **Python 版本** | 3.8+ | 3.10 或 3.11 或 3.12 |
| **GPU 型号** | NVIDIA GPU (计算能力 7.0+) | T4 / V100 / RTX 3080 / RTX 4090 |
| **GPU 显存** | ≥ 12 GB | ≥ 16 GB |
| **系统内存** | ≥ 12 GB RAM | ≥ 16 GB RAM |
| **存储空间** | ≥ 20 GB 可用空间 | ≥ 50 GB SSD |
| **CUDA 版本** | 11.8+ | 12.1+ |
| **cuDNN 版本** | 8.0+ | 8.9+ |

💡 **GPU 说明**：
- ✅ 支持的 GPU：GTX 1080Ti, RTX 2080/3060/3070/3080/3090/4090, T4, V100, A10, A100
- ⚠️ 低于 12GB 显存可能导致 OOM（内存溢出）错误
- 📊 T4 16GB 是性价比最高的选择（云服务器推荐）
- 🚫 不支持 CPU 训练（速度太慢，不推荐）

#### Python 依赖包版本（经过验证）

```bash
# 核心深度学习框架
torch==2.9.0+cu126           # PyTorch（CUDA 12.6）
transformers==4.57.2         # Hugging Face 模型库
peft==0.18.0                 # 参数高效微调核心库

# 训练加速与工具
accelerate==1.12.0           # 分布式训练加速
safetensors==0.7.0           # 安全的模型权重格式

# 数据处理与可视化
pandas==2.2.2                # 数据处理
numpy==2.0.2                 # 数值计算
tqdm==4.67.1                 # 进度条显示
matplotlib==3.10.0           # 数据可视化

# 网络请求
requests==2.32.4             # HTTP 请求库
```

💡 **版本兼容性说明**：
- 上述版本已在 Google Colab 和本地环境测试通过
- PyTorch 版本需与 CUDA 版本匹配：
  - CUDA 11.8: `torch==2.0.0+cu118`
  - CUDA 12.1: `torch==2.1.0+cu121`
  - CUDA 12.6: `torch==2.9.0+cu126`
- ⚠️ **重要**：transformers 和 peft 版本需保持兼容，建议使用上述版本组合

---


# 1️⃣ 环境准备和依赖安装

## 📦 依赖安装说明

### 🌐 方式 1：Google Colab（推荐）

Google Colab 已预装部分深度学习依赖，只需运行下方单元格安装/升级少量包即可。

#### 依赖包版本说明

| 依赖库 | 版本 | 作用 | 安装说明 |
|:------|:-----|:----|:---------|
| **torch** | 2.9.0+cu126 | PyTorch 深度学习框架 | Colab 预装，可选升级 |
| **transformers** | 4.57.2 | Hugging Face 预训练模型库 | 建议升级到此版本 |
| **peft** | 0.18.0 | 参数高效微调核心库 | 必需安装 |
| **accelerate** | 1.12.0 | 分布式训练加速库 | 建议升级 |
| **pandas** | 2.2.2 | 结构化数据处理 | Colab 预装 |
| **numpy** | 2.0.2 | 数值计算基础库 | Colab 预装 |
| **tqdm** | 4.67.1 | 进度条组件 | Colab 预装 |
| **matplotlib** | 3.10.0 | 数据可视化库 | Colab 预装 |
| **requests** | 2.32.4 | HTTP 请求库 | Colab 预装 |
| **safetensors** | 0.7.0 | 安全模型权重格式 | 建议升级 |

💡 **安装说明**：
- ✅ 使用 `-q` 参数静默安装，减少输出信息
- ✅ 使用 `--upgrade` 确保安装最新版本
- ⚠️ 如果在 Colab 中运行，预装包会自动跳过

### 🏠 方式 2：私有环境

如果在本地或私有云环境运行，请先安装 PyTorch：

```bash
# 第一步：根据 CUDA 版本安装 PyTorch
# CUDA 11.8
pip install torch==2.0.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

# CUDA 12.1
pip install torch==2.1.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

# CUDA 12.6 (推荐，与 Colab 一致)
pip install torch==2.9.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/cu126

# 第二步：安装其他依赖（运行下方单元格即可）
```

💡 **注意事项**：
- 🔍 先运行 `nvidia-smi` 查看 CUDA 版本
- 🌐 如果网络较慢，可以使用镜像源：`pip install -i https://pypi.tuna.tsinghua.edu.cn/simple`
- 💾 建议使用虚拟环境：`python -m venv lora_env && source lora_env/bin/activate`

In [ ]:
# 📦 安装依赖包
# 说明：使用 -q 参数安静模式安装，--upgrade 确保安装最新版本
# 如果在 Colab 中运行，某些包可能已预装，会自动跳过

%pip install -q --upgrade \
    torch==2.9.0+cu126  \
    transformers==4.57.2 \
    peft==0.18.0 \
    accelerate==1.12.0 \
    pandas==2.2.2 \
    numpy==2.0.2 \
    tqdm==4.67.1 \
    matplotlib==3.10.0 \
    requests==2.32.4 \
    safetensors==0.7.0

print("✅ 依赖包安装完成！")

✅ 依赖包安装完成！


# 2️⃣ 导入依赖库和环境检查

## 🔧 核心库导入说明

下面我们将导入所有需要的库，并检查运行环境。


In [ ]:
# 🔧 导入所有必需的库
# 功能：准备训练所需的所有依赖

import io
import os
import random
import zipfile

import pandas as pd
import torch
from torch.utils.data import Dataset, DataLoader
from transformers import (
    AutoModelForSequenceClassification,  # 序列分类模型（用于文本分类）
    AutoTokenizer,                        # 自动选择合适的分词器
    get_linear_schedule_with_warmup,     # 带预热的线性学习率调度器
)
from peft import (
    LoraConfig,        # LoRA 配置类
    TaskType,          # 任务类型枚举
    get_peft_model,    # 将 LoRA 应用到模型的核心函数
)
from tqdm import tqdm  # 进度条显示
import requests        # HTTP 请求（用于下载数据）

# 🎯 设置随机种子，确保结果可复现
# 说明：在深度学习中，很多操作涉及随机性（如参数初始化、数据打乱）
#      固定随机种子可以让每次运行得到相同的结果，便于调试和对比
torch.manual_seed(42)   # PyTorch 的随机种子
random.seed(42)         # Python 标准库的随机种子

# 🎮 设备选择：优先使用 GPU，如果没有则使用 CPU
# 说明：CUDA 是 NVIDIA GPU 的并行计算平台
#      torch.cuda.is_available() 检查是否有可用的 GPU
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# 📊 显示环境信息
print("🔍 环境检查")
print("=" * 60)
print(f"🎮 使用设备: {device}")
if torch.cuda.is_available():
    print(f"   GPU 型号: {torch.cuda.get_device_name(0)}")
    print(f"   GPU 数量: {torch.cuda.device_count()}")
    print(f"   CUDA 版本: {torch.version.cuda}")
else:
    print("   ⚠️  未检测到 GPU，将使用 CPU（训练速度较慢）")

print(f"🐍 Python 库版本:")
print(f"   PyTorch: {torch.__version__}")
try:
    import transformers
    print(f"   Transformers: {transformers.__version__}")
    import peft
    print(f"   PEFT: {peft.__version__}")
except ImportError as e:
    print(f"   ⚠️ 部分库未安装: {e}")

print("\n✅ 环境准备完成！")


🔍 环境检查
🎮 使用设备: cuda
   GPU 型号: Tesla T4
   GPU 数量: 1
   CUDA 版本: 12.6
🐍 Python 库版本:
   PyTorch: 2.9.0+cu126
   Transformers: 4.57.2
   PEFT: 0.18.0

✅ 环境准备完成！


# 3️⃣ 数据准备和预处理

## 📊 数据集介绍

我们使用 **SMS Spam Collection** 数据集进行垃圾短信分类任务：

- **数据来源**：UCI Machine Learning Repository
- **数据规模**：约 5,572 条短信
- **任务类型**：二分类（ham 正常短信 vs spam 垃圾短信）
- **标签分布**：不平衡（ham 约 87%, spam 约 13%）

## 🔧 数据处理流程

```
原始数据下载 → 类别平衡采样 → 划分训练/验证/测试集 → 分词编码 → DataLoader加载
```


In [ ]:
# 📊 数据处理工具函数
# 功能：提供数据下载、平衡、划分和加载的完整工具集

def create_balanced_dataset(df):
    """
    创建类别平衡的数据集

    🎯 目的：解决数据不平衡问题
    原始数据中 ham（正常短信）远多于 spam（垃圾短信），
    这会导致模型偏向预测多数类。通过下采样多数类，
    使两个类别的样本数量相等。

    参数：
        df: pandas DataFrame，包含 Label 和 Text 列

    返回：
        balanced_df: 类别平衡后的 DataFrame

    示例：
        原始：ham(4825条) + spam(747条) = 5572条
        平衡后：ham(747条) + spam(747条) = 1494条
    """
    label_col = df["Label"]

    # 🔍 兼容性处理：标签可能是字符串（'spam'/'ham'）或数字（0/1）
    spam_mask = (label_col == "spam") | (label_col == 1)
    ham_mask = (label_col == "ham") | (label_col == 0)

    # 统计垃圾短信数量（少数类）
    spam_count = int(spam_mask.sum())

    if spam_count == 0 or ham_mask.sum() == 0:
        raise ValueError("数据集缺少 spam 或 ham 类别，无法平衡")

    # 🎲 从多数类（ham）中随机采样，数量与少数类相同
    # random_state=123 确保每次采样结果相同，便于复现
    ham_subset = df[ham_mask].sample(spam_count, random_state=123)

    # 📦 合并两个类别的数据
    balanced = pd.concat([ham_subset, df[spam_mask]])

    # 🔀 随机打乱数据，避免数据顺序偏差
    return balanced.sample(frac=1, random_state=123).reset_index(drop=True)


def random_split(df, train_frac=0.7, val_frac=0.1):
    """
    随机划分数据集为训练集、验证集和测试集

    🎯 目的：将数据分为三部分
    - 训练集：用于模型参数更新
    - 验证集：用于调整超参数和早停
    - 测试集：用于最终性能评估

    参数：
        df: pandas DataFrame
        train_frac: 训练集比例（默认 0.7 = 70%）
        val_frac: 验证集比例（默认 0.1 = 10%）

    返回：
        train_df, val_df, test_df: 三个数据框

    示例：
        总共1494条 → 训练1046条(70%) + 验证149条(10%) + 测试299条(20%)
    """
    # 🔀 随机打乱数据
    df = df.sample(frac=1, random_state=123).reset_index(drop=True)

    # 📏 计算划分点
    train_end = int(len(df) * train_frac)
    val_end = train_end + int(len(df) * val_frac)

    # ✂️ 执行划分
    train_df = df[:train_end]
    val_df = df[train_end:val_end]
    test_df = df[val_end:]

    return train_df, val_df, test_df


def ensure_sms_file(local_path="./SMSSpamCollection.tsv"):
    """
    确保 SMS 数据文件存在，如果不存在则自动下载

    🎯 目的：自动化数据准备流程
    如果本地已有数据文件，直接返回路径；
    如果没有，则从 UCI 仓库下载并解压。

    参数：
        local_path: 本地文件路径（可选）

    返回：
        str: 数据文件的实际路径

    说明：
        数据文件格式为 TSV（Tab-Separated Values），每行包含标签和文本
    """
    # ✅ 如果文件已存在，直接返回
    if os.path.exists(local_path):
        return local_path

    # 📥 从 UCI 机器学习仓库下载数据
    url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip"
    print(f"📥 正在从 UCI 仓库下载数据...")
    print(f"   URL: {url}")

    try:
        # 🌐 发送 HTTP 请求下载 ZIP 文件
        resp = requests.get(url, timeout=30)
        resp.raise_for_status()  # 如果请求失败，抛出异常

        # 📦 解压 ZIP 文件，读取其中的 SMSSpamCollection 文件
        with zipfile.ZipFile(io.BytesIO(resp.content)) as zf:
            raw = zf.read("SMSSpamCollection").decode("utf-8")

        # 💾 保存到本地文件
        out_path = "SMSSpamCollection.tsv"
        with open(out_path, "w", encoding="utf-8") as f:
            f.write(raw)

        print(f"✅ 数据下载成功，保存到: {out_path}")
        return out_path

    except Exception as e:
        print(f"❌ 数据下载失败: {e}")
        print(f"💡 请手动下载并放置到当前目录")
        raise


def load_sms_dataframe():
    """
    加载并预处理 SMS 数据集

    🎯 目的：一站式数据加载函数
    整合了下载、读取、平衡、标签转换和划分的完整流程

    返回：
        train_df, val_df, test_df: 三个已处理的 DataFrame

    数据处理步骤：
        1. 确保数据文件存在（自动下载）
        2. 读取 TSV 文件
        3. 平衡类别分布
        4. 标签转换（ham→0, spam→1）
        5. 划分训练/验证/测试集
    """
    # 📂 获取数据文件路径（如需要会自动下载）
    data_path = ensure_sms_file()

    # 📖 读取 TSV 文件
    # sep="\t" 指定制表符分隔
    # names 指定列名
    # header=None 表示文件没有表头行
    df = pd.read_csv(data_path, sep="\t", names=["Label", "Text"], header=None)

    print(f"📊 原始数据加载完成: {len(df)} 条")
    print(f"   - ham (正常): {(df['Label']=='ham').sum()} 条")
    print(f"   - spam (垃圾): {(df['Label']=='spam').sum()} 条")

    # ⚖️ 创建类别平衡的数据集
    balanced = create_balanced_dataset(df)
    print(f"⚖️  数据平衡后: {len(balanced)} 条（每类 {len(balanced)//2} 条）")

    # 🔄 标签转换：字符串 → 数字
    # 机器学习模型需要数字标签
    label_map = {"ham": 0, "spam": 1}
    balanced["Label"] = balanced["Label"].apply(
        lambda v: label_map[v] if v in label_map else int(v)
    )

    # ✂️ 划分数据集
    train_df, val_df, test_df = random_split(balanced, train_frac=0.7, val_frac=0.1)
    print(f"✂️  数据划分完成:")
    print(f"   - 训练集: {len(train_df)} 条")
    print(f"   - 验证集: {len(val_df)} 条")
    print(f"   - 测试集: {len(test_df)} 条")

    return train_df, val_df, test_df


# 🎓 初学者提示：
# 以上函数构成了完整的数据处理流水线，每个函数职责单一，便于理解和调试。
# 在实际项目中，建议将数据处理代码模块化，提高代码复用性。
print("✅ 数据处理函数定义完成")


✅ 数据处理函数定义完成


In [ ]:
# 📦 PyTorch Dataset 类定义
# 功能：将 DataFrame 转换为 PyTorch 可用的数据集格式

class SpamSequenceDataset(Dataset):
    """
    SMS 垃圾短信数据集类

    🎯 目的：将 pandas DataFrame 转换为 PyTorch Dataset
    这个类继承自 torch.utils.data.Dataset，实现了 PyTorch 数据加载的标准接口

    与手动实现的区别：
    - 手动实现：使用 tiktoken 分词器，手动处理 padding
    - PEFT版本：使用 Hugging Face Tokenizer，自动处理 padding 和 attention_mask

    关键特性：
    1. 使用 Hugging Face tokenizer，与模型完美配合
    2. 自动生成 attention_mask（标记哪些是真实 token，哪些是 padding）
    3. 批量预处理所有文本，提高效率

    SMS 垃圾短信数据集格式示例(每一行由两列组成：第一列是标签（ham 表示正常信息，spam 表示垃圾信息），第二列是原始文本)：
    ham   What you doing?how are you?
    ham   Ok lar... Joking wif u oni...
    ham   dun say so early hor... U c already then say...
    ham   MY NO. IN LUTON 0125698789 RING ME IF UR AROUND! H*
    ham   Siva is in hostel aha:-.
    ham   Cos i was out shopping wif darren jus now n i called him 2 ask wat present he wan lor. Then he started guessing who i was wif n he finally guessed darren lor.
    spam   FreeMsg: Txt: CALL to No: 86888 & claim your reward of 3 hours talk time to use from your phone now! ubscribe6GBP/ mnth inc 3hrs 16 stop?txtStop
    spam   Sunshine Quiz! Win a super Sony DVD recorder if you canname the capital of Australia? Text MQUIZ to 82277. B
    spam   URGENT! Your Mobile No 07808726822 was awarded a L2,000 Bonus Caller Prize on 02/09/03! This is our 2nd attempt to contact YOU! Call 0871-872-9758 BOX95QU
    """

    def __init__(self, df, tokenizer, max_length=96):
        """
        初始化数据集

        参数：
            df: pandas DataFrame，包含 'Text' 和 'Label' 列
            tokenizer: Hugging Face tokenizer 对象
            max_length: 最大序列长度，超过会截断，不足会填充

        处理流程：
            1. 批量分词所有文本
            2. 统一长度（截断/填充到 max_length）
            3. 转换为 PyTorch tensors
            4. 保存 labels 为数字张量
        """
        # 🔤 批量分词所有文本
        # 使用 Hugging Face Tokenizer 对所有文本进行编码
        # tokenizer() 函数的参数说明：
        # - df["Text"].tolist(): 将 DataFrame 的 'Text' 列转换为 Python 列表作为输入。
        # - truncation=True: 启用截断。如果文本长度超过 `max_length`，则将其截断。
        # - padding="max_length": 启用填充。将所有序列填充到 `max_length` 指定的长度。
        # - max_length=96: 指定所有序列的最终长度。较短的序列将被填充，较长的序列将被截断。
        # - return_tensors="pt": 返回 PyTorch 张量格式的编码结果，方便后续模型使用。
        encodings = tokenizer(
            df["Text"].tolist(),          # 将 pandas Series 转为 list，供分词器处理
            truncation=True,               # 启用截断，保证序列长度不超过 max_length
            padding="max_length",          # 填充到最大长度，使得批次内的所有序列长度一致
            max_length=max_length,         # 最大序列长度，定义了输入模型文本的最大token数
            return_tensors="pt",           # 返回 PyTorch 张量格式，方便后续在PyTorch中操作
        )

        # 📊 保存编码结果
        # input_ids: 表示每个文本对应的 token ID 序列，形状为 [N, max_length]
        # 其中 N 是数据集中的样本数量。
        self.input_ids = encodings["input_ids"]

        # attention_mask: 注意力掩码，形状为 [N, max_length]
        # 值为 1 的位置表示是原始文本的有效 token，值为 0 的位置表示是填充（padding）token。
        # 模型在计算注意力时会忽略 attention_mask=0 的位置，避免填充对模型决策的影响。
        self.attention_mask = encodings["attention_mask"]

        # 🏷️ 保存标签
        # 将 DataFrame 的 'Label' 列转换为 Python 列表，再转换为 PyTorch 张量。
        # 转换为 torch.long 类型（即 int64），因为 PyTorch 的交叉熵损失函数通常要求标签为 long 类型。
        self.labels = torch.tensor(df["Label"].tolist(), dtype=torch.long)

        # 📏 保存数据集大小
        self._len = len(self.labels)

        print(f"✅ 数据集创建完成:")
        print(f"   - 样本数: {self._len}")
        print(f"   - 序列长度: {max_length}")
        print(f"   - 输入形状: {self.input_ids.shape}")

    def __len__(self):
        """
        返回数据集大小

        🎯 目的：告诉 DataLoader 数据集有多少样本
        这是 PyTorch Dataset 必须实现的方法之一
        """
        return self._len

    def __getitem__(self, idx):
        """
        获取单个样本

        🎯 目的：根据索引返回一个样本
        这是 PyTorch Dataset 必须实现的方法之一

        参数：
            idx: 样本索引（0 到 len-1）

        返回：
            dict: 包含 input_ids, attention_mask, labels 的字典

        说明：
            返回字典格式是 Hugging Face 模型的标准输入格式，
            可以直接使用 model(**batch) 进行解包传参
        """
        return {
            "input_ids": self.input_ids[idx],           # Token ID 序列
            "attention_mask": self.attention_mask[idx], # 注意力掩码
            "labels": self.labels[idx],                 # 标签（0 或 1）
        }


# 🎓 初学者提示：
# Dataset 类只负责数据的"组织"，不负责数据的"批量加载"。
# 批量加载由 DataLoader 完成，它会：
# 1. 调用 __getitem__ 获取多个样本
# 2. 自动将这些样本组合成 batch
# 3. 可选地打乱数据顺序（shuffle=True）
# 4. 可选地使用多进程加速（num_workers>0）

print("✅ Dataset 类定义完成")

✅ Dataset 类定义完成


# 4️⃣ 执行数据加载

## 🚀 创建 Tokenizer 和加载数据
Tokenizer 的核心作用是把原始的文本（一长串字符）分割成模型可以理解的、更小的单元，这些单元被称为 tokens。这个过程就叫做 tokenization (分词)。

代码中使用的 `AutoTokenizer.from_pretrained("gpt2")` 就是加载了一个预训练的 GPT-2 分词器，它会将您的输入文本转换为 GPT-2 模型能够处理的 token ID 序列，然后这些 ID 会在模型内部被转换成词嵌入。

现在我们将：
1. 加载 GPT-2 tokenizer
2. 加载并处理 SMS 数据集
3. 创建 PyTorch DataLoader


In [ ]:
# 🔤 加载 Tokenizer
# 功能：从 Hugging Face Hub 加载预训练的 GPT-2 分词器

print("🔤 加载 GPT-2 Tokenizer...")

# 📥 从 Hugging Face 下载并加载 GPT-2 tokenizer
# AutoTokenizer.from_pretrained() 会：
# 1. 检查本地缓存（~/.cache/huggingface/）
# 2. 如果没有，从 Hub 下载
# 3. 加载词表和分词规则
tokenizer = AutoTokenizer.from_pretrained("gpt2")

# 🔧 配置 padding token
# GPT-2 原始设计用于生成任务，没有 padding token
# 但分类任务需要将不同长度的文本对齐到相同长度
# 解决方案：将 eos_token（结束符）复用为 pad_token
if tokenizer.pad_token_id is None:
    tokenizer.pad_token = tokenizer.eos_token
    print(f"   ⚙️  设置 pad_token = eos_token (ID: {tokenizer.eos_token_id})")

print(f"✅ Tokenizer 加载完成")
print(f"   - 词表大小: {tokenizer.vocab_size:,}")
print(f"   - 模型类型: {tokenizer.__class__.__name__}")

# 📊 加载并划分数据
print("\n📊 加载 SMS 数据集...")
train_df, val_df, test_df = load_sms_dataframe()

# 📏 设置最大序列长度
# 说明：这个值需要平衡以下因素：
# - 太短：可能截断重要信息
# - 太长：增加计算成本和显存占用
# - 96 是经过实验的合理值，足以覆盖大部分短信
max_length = 96
print(f"\n📏 设置最大序列长度: {max_length}")

# 🔨 创建 PyTorch Dataset
print("\n🔨 创建 PyTorch Dataset...")
train_dataset = SpamSequenceDataset(train_df, tokenizer, max_length=max_length)
val_dataset = SpamSequenceDataset(val_df, tokenizer, max_length=max_length)
test_dataset = SpamSequenceDataset(test_df, tokenizer, max_length=max_length)

# 📦 创建 DataLoader
# 功能：批量加载数据，支持自动打乱、多进程等
print("\n📦 创建 DataLoader...")

# 🎯 DataLoader 参数说明：
# - batch_size: 每批样本数量
#   - 较大：训练快，但显存占用高
#   - 较小：显存友好，但训练慢
#   - 8 是适中的值
# - shuffle: 是否打乱数据
#   - 训练集：True（避免顺序偏差）
#   - 验证/测试集：False（保持顺序，便于分析）

train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=8, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=8, shuffle=False)

print(f"✅ DataLoader 创建完成")
print(f"   - 训练批次数: {len(train_loader)}")
print(f"   - 验证批次数: {len(val_loader)}")
print(f"   - 测试批次数: {len(test_loader)}")
print(f"   - 每批大小: 8")

# 🔍 数据检查：查看一个样本
print("\n🔍 样本检查:")
sample_batch = next(iter(train_loader))
print(f"   - input_ids 形状: {sample_batch['input_ids'].shape}")
print(f"   - attention_mask 形状: {sample_batch['attention_mask'].shape}")
print(f"   - labels 形状: {sample_batch['labels'].shape}")
print(f"   - 标签分布: {sample_batch['labels'].tolist()}")

print("\n✅ 数据准备完成！")


🔤 加载 GPT-2 Tokenizer...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

   ⚙️  设置 pad_token = eos_token (ID: 50256)
✅ Tokenizer 加载完成
   - 词表大小: 50,257
   - 模型类型: GPT2TokenizerFast

📊 加载 SMS 数据集...
📥 正在从 UCI 仓库下载数据...
   URL: https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
✅ 数据下载成功，保存到: SMSSpamCollection.tsv
📊 原始数据加载完成: 5572 条
   - ham (正常): 4825 条
   - spam (垃圾): 747 条
⚖️  数据平衡后: 1494 条（每类 747 条）
✂️  数据划分完成:
   - 训练集: 1045 条
   - 验证集: 149 条
   - 测试集: 300 条

📏 设置最大序列长度: 96

🔨 创建 PyTorch Dataset...
✅ 数据集创建完成:
   - 样本数: 1045
   - 序列长度: 96
   - 输入形状: torch.Size([1045, 96])
✅ 数据集创建完成:
   - 样本数: 149
   - 序列长度: 96
   - 输入形状: torch.Size([149, 96])
✅ 数据集创建完成:
   - 样本数: 300
   - 序列长度: 96
   - 输入形状: torch.Size([300, 96])

📦 创建 DataLoader...
✅ DataLoader 创建完成
   - 训练批次数: 131
   - 验证批次数: 19
   - 测试批次数: 38
   - 每批大小: 8

🔍 样本检查:
   - input_ids 形状: torch.Size([8, 96])
   - attention_mask 形状: torch.Size([8, 96])
   - labels 形状: torch.Size([8])
   - 标签分布: [0, 0, 0, 1, 1, 0, 0, 1]

✅ 数据准备完成！


### 数据集中的“形状”和“标签分布”代表什么：

想象一下你的数据就像一堆短信，每次不是只给模型一条短信，而是给一小“批”短信，这样模型可以更高效地学习。
- **`input_ids` 形状: `torch.Size([8, 96])`**

  - `8`：这代表你这一批次（batch）有 **8 条短信**。你的 DataLoader（数据加载器）一次性打包了 8 条短信给模型处理。
  - `96`：这代表每一条短信都被处理成了 **96 个“数字编码”**。如果短信本身不够 96 个编码，就会在后面补齐（填充）；如果短信太长，就会被截断到 96 个编码。所以，每条短信现在都是一个固定长度的数字序列。
  - **通俗理解**：你一次性拿出了 8 张表格，每张表格有 96 个格子，每个格子里面都填了一个数字（代表短信里的一个词或词的一部分）。

- **`attention_mask` 形状: `torch.Size([8, 96])`**

  - 同样，`8` 代表 **8 条短信**，`96` 代表每条短信的 **96 个位置**。
  - 注意力掩码会告诉模型，在这 96 个位置中，哪些是短信“真正”的内容（值为 1），哪些是后面为了凑齐 96 个位置而补上的“空白”（值为 0）。
  - **通俗理解**：你还有 8 张表格，每张表格也是 96 个格子。这次格子里的数字不是短信内容，而是标记：如果是短信的真内容就打勾（1），是填充的空白就打叉（0）。这样模型就知道哪些信息是重要的。

- **`labels` 形状: `torch.Size([8])`**

  - 这代表你有 **8 个标签**，每个标签对应这一批次中的一条短信。
  - **通俗理解**：你有一张小纸条，上面写了 8 个数字，每个数字都对应着前面那 8 条短信的分类结果（比如第一条短信是“正常”，第二条是“垃圾”）。

- **标签分布: `[0, 0, 0, 1, 1, 0, 0, 1]`**

  - 这是那 8 个标签的具体内容。

  - 通常，`0` 代表“正常短信 (ham)”；`1` 代表“垃圾短信 (spam)”。

  - **通俗理解**：在你这张小纸条上，写着：

    - 第 1 条短信：0（正常）
    - 第 2 条短信：0（正常）
    - 第 3 条短信：0（正常）
    - 第 4 条短信：1（垃圾）
    - 第 5 条短信：1（垃圾）
    - 第 6 条短信：0（正常）
    - 第 7 条短信：0（正常）
    - 第 8 条短信：1（垃圾）

    所以，这一批 8 条短信里有 5 条正常短信和 3 条垃圾短信。

# 5️⃣ 使用 PEFT 库应用 LoRA

## 🎯 PEFT 核心概念

**PEFT (Parameter-Efficient Fine-Tuning)** 是 Hugging Face 提供的参数高效微调库，支持多种方法：

- **LoRA**: Low-Rank Adaptation（本教程使用）
- **LoHa**: Low-Rank Hadamard Product
- **LoKr**: Low-Rank Kronecker Product
- **Prefix Tuning**: 可学习的前缀
- **P-Tuning**: 提示微调
- **Adapter**: 适配器层

## 🔧 LoRA 配置详解

使用 PEFT 库应用 LoRA 只需三步：
1. 创建 `LoraConfig` 配置对象
2. 加载基础模型
3. 使用 `get_peft_model()` 应用 LoRA

让我们深入了解每个参数的含义。


In [ ]:
# 🤖 加载基础模型并应用 LoRA
# 功能：这是 PEFT 库使用的核心步骤

print("🤖 步骤 1: 加载基础模型...")
print("=" * 60)

# 📥 从 Hugging Face Hub 加载 GPT-2 分类模型
# AutoModelForSequenceClassification 会：
# 1. 加载 GPT-2 的预训练权重
# 2. 在顶部添加一个分类头（线性层）
# 3. num_labels=2 表示二分类任务
base_model = AutoModelForSequenceClassification.from_pretrained(
    "gpt2",           # 模型名称
    num_labels=2      # 分类类别数
)

# 🔧 配置模型的 padding token
# 必须与 tokenizer 的配置保持一致
base_model.config.pad_token_id = tokenizer.pad_token_id

# 📏 调整 embedding 层大小
# GPT-2 的词表大小是 50257，如果 tokenizer 添加了新 token，需要调整
base_model.resize_token_embeddings(len(tokenizer))

print(f"✅ 基础模型加载完成")
print(f"   - 模型类型: {base_model.__class__.__name__}")
print(f"   - 分类类别: {base_model.config.num_labels}")
print(f"   - 词表大小: {base_model.config.vocab_size:,}")

# 🔧 步骤 2: 创建 LoRA 配置
print("\n🔧 步骤 2: 创建 LoRA 配置...")
print("=" * 60)

# 📋 LoRA 配置参数详解
lora_config = LoraConfig(
    # 🎯 task_type: 任务类型
    # - TaskType.SEQ_CLS: 序列分类任务
    # - TaskType.CAUSAL_LM: 因果语言模型（生成任务）
    # - TaskType.SEQ_2_SEQ_LM: 序列到序列任务
    # 不同任务类型会影响 LoRA 层的插入位置
    task_type=TaskType.SEQ_CLS,

    # 📊 r: LoRA 的秩（rank）
    # 这是 LoRA 最核心的超参数！
    # - 控制 A 和 B 矩阵的中间维度
    # - 较小的值（如 4-8）：参数少，训练快，但表达能力弱
    # - 较大的值（如 16-32）：参数多，表达能力强，但接近全量微调
    # - 推荐起始值：8
    # - 本教程使用：8（平衡效率和性能）
    r=8,

    # 🎚️ lora_alpha: LoRA 缩放因子
    # - 控制 LoRA 输出的影响强度
    # - 常见设置：alpha = 2×r 或 alpha = r
    # - 本教程：alpha = 2×r = 16
    # - 数学公式：LoRA 输出会乘以 (alpha/r) 的缩放系数
    lora_alpha=16,

    # 🎲 lora_dropout: Dropout 比率
    # lora_dropout 就是训练LoRA时随机让部分参数“休息”，
    #        避免模型死记硬背、提高泛化能力的正则手段——适当加一点更聪明，太高反而学不会。
    # - 在 LoRA 层中应用 dropout 正则化
    # - 范围：0.0（不使用）到 0.2（较强正则化）
    # - 作用：防止过拟合，提高泛化能力
    # - 本教程：0.05（轻微正则化）
    lora_dropout=0.05,

    # 🎯 target_modules: 要替换的模块名称列表
    # 这是 PEFT 库的强大功能：选择性地只对部分层应用 LoRA
    # GPT-2 的模块名称：
    # - "c_attn": 注意力层的 QKV 投影（最重要）
    # - "c_proj": 注意力层的输出投影
    # - "c_fc": 前馈网络的第一层
    # - "c_proj": 前馈网络的第二层（与注意力同名，PEFT 会区分）
    #
    # 💡 选择策略：
    # - 只替换 ["c_attn"]: 最少参数，适合显存极小场景
    # - 替换 ["c_attn", "c_proj"]: 平衡选择
    # - 替换 ["c_attn", "c_fc", "c_proj"]: 本教程使用，覆盖主要层
    # - 不建议替换 embedding 层和分类头
    ## 给 GPT-2 的注意力 QKV、注意力输出、MLP 前后两层都加上 LoRA 小“外挂”，
    ## 只改这些 Linear，就能让模型的“注意力习惯”和“特征变换方式”适配新任务。
    target_modules=["c_attn", "c_fc", "c_proj"],

    # ⚙️ 其他可选参数（使用默认值）：
    # - bias="none": 不训练 bias 参数
    # - fan_in_fan_out=False: 权重矩阵的组织方式
    # - modules_to_save=None: 除 LoRA 外需要训练的模块
)

print("✅ LoRA 配置创建完成:")
print(f"   - 秩 (r): {lora_config.r}")
print(f"   - 缩放因子 (alpha): {lora_config.lora_alpha}")
print(f"   - Dropout: {lora_config.lora_dropout}")
print(f"   - 目标模块: {lora_config.target_modules}")

# 🚀 步骤 3: 应用 LoRA 到模型
print("\n🚀 步骤 3: 应用 LoRA...")
print("=" * 60)

# 🔑 核心函数：get_peft_model()
# 功能：将 LoRA 配置应用到基础模型
# 工作原理：
# 1. 遍历模型的所有模块
# 2. 找到 target_modules 中指定的线性层
# 3. 用 LoRA 增强版本替换这些层
# 4. 冻结原始参数，只让 LoRA 参数可训练
model = get_peft_model(base_model, lora_config)

print("✅ LoRA 应用完成！")

# 📊 打印参数统计
# print_trainable_parameters() 是 PEFT 模型的便捷方法
# 会显示可训练参数数量和占比
print("\n📊 参数统计:")
model.print_trainable_parameters()

# 🎮 将模型移动到 GPU/CPU
model.to(device)
print(f"\n🎮 模型已移动到设备: {device}")

# 🔍 模型结构检查（可选）
print("\n🔍 模型结构预览:")
print(f"   - 模型类型: {type(model).__name__}")
print(f"   - 基础模型: {type(model.base_model).__name__}")
print(f"   - LoRA 配置: {model.peft_config}")

print("\n✅ 模型准备完成！")


🤖 步骤 1: 加载基础模型...


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

Some weights of GPT2ForSequenceClassification were not initialized from the model checkpoint at gpt2 and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.12/dist-packages/peft/tuners/lora/layer.py:2285: UserWarning: fan_in_fan_out is set to False but the target module is `Conv1D`. Setting fan_in_fan_out to True.
  warnings.warn(


✅ 基础模型加载完成
   - 模型类型: GPT2ForSequenceClassification
   - 分类类别: 2
   - 词表大小: 50,257

🔧 步骤 2: 创建 LoRA 配置...
✅ LoRA 配置创建完成:
   - 秩 (r): 8
   - 缩放因子 (alpha): 16
   - Dropout: 0.05
   - 目标模块: {'c_proj', 'c_fc', 'c_attn'}

🚀 步骤 3: 应用 LoRA...
✅ LoRA 应用完成！

📊 参数统计:
trainable params: 1,181,184 || all params: 125,622,528 || trainable%: 0.9403

🎮 模型已移动到设备: cuda

🔍 模型结构预览:
   - 模型类型: PeftModelForSequenceClassification
   - 基础模型: LoraModel
   - LoRA 配置: {'default': LoraConfig(task_type=<TaskType.SEQ_CLS: 'SEQ_CLS'>, peft_type=<PeftType.LORA: 'LORA'>, auto_mapping=None, peft_version='0.18.0', base_model_name_or_path='gpt2', revision=None, inference_mode=False, r=8, target_modules={'c_proj', 'c_fc', 'c_attn'}, exclude_modules=None, lora_alpha=16, lora_dropout=0.05, fan_in_fan_out=True, bias='none', use_rslora=False, modules_to_save=['classifier', 'score'], init_lora_weights=True, layers_to_transform=None, layers_pattern=None, rank_pattern={}, alpha_pattern={}, megatron_config=None, megatron_core='meg

### 🔍 LoRA 应用结果解释

| 结果项                                    | 通俗含义                                          |
| ----------------------------------------- | ------------------------------------------------- |
| `trainable params: 1,181,184`             | 只训练约 118 万参数，非常轻量                     |
| `all params: 125,622,528`                 | GPT-2 原始参数约 1.25 亿                          |
| `trainable%: 0.9403`                      | 只动 **0.94%** 的参数，99% 不改动                 |
| `cuda`                                    | 模型成功上 GPU，训练可以开跑                      |
| 类型 `PeftModelForSequenceClassification` | 你的 GPT-2 现在用于分类任务                       |
| `LoRA 配置` 里 `c_attn, c_fc, c_proj`     | LoRA 安装在注意力与前馈核心层，最常用最有效       |
| `lora_dropout=0.05`                       | 防止过拟合的小保护，不让模型“背题”                |
| `modules_to_save=['classifier','score']`  | 除了 LoRA，你还保留训练分类头层，适合下游分类任务 |

一句话总结：

> **你不是在“重训 GPT-2”，而是在它脑子里挂了个小外挂让它学会新任务。**

### GPT-2的词表和向量维度

GPT-2 的词表大小是 50257，意思是它能用 **50257 个基本符号（token）** 来表示所有文本，是英文里最常见字词 + 子词 + 标点组合成的一套“拆词积木”。

GPT-2 的向量维度是 768。每个词表都有 一个 768 维参数描述（embedding 向量）

#### 一张图，彻底理解 50257 与 768 的关系

```
                ┌───────────────────────────────┐
                │         GPT-2 词表 (Vocab)    │
                │     大小 = 50257 个 Token     │
                └───────────────────────────────┘
                         │       │       │
                         │       │       │  ... 共有 50257 条
                         ▼       ▼       ▼
                    Token#1  Token#2  Token#3             Token#50257
                         │       │       │                     │
                         │       │       │                     │
                         ▼       ▼       ▼                     ▼
          ┌──────────────────────────────────────────────────────────────┐
          │           Token Embedding Matrix (查词典用的矩阵)             │
          │          形状 = 50257 × 768  （行=词表数，列=向量维度）        │
          │  ┌────────────────────────────────────────────────────────┐  │
   Row#1  -> │  [ 0.13 , -0.88 , ... , 0.04 ]  ← 768维                       │
   Row#2  -> │  [ 0.51 ,  1.22 , ... , -0.91 ] ← 768维                       │
   Row#3  -> │  [ -0.07,  0.33 , ... , 1.85 ]  ← 768维                       │
    ...    ->│                                                              │
   Row50257->│  [ 0.09 , -0.44 , ... , 0.13 ]  ← 768维                       │
          └──────────────────────────────────────────────────────────────┘
```

------

#### 用图一句话总结

> **50257 是行数（有多少 Token）**
>  **768 是列数（每个 Token 向量有多长）**

🔸 Token 多 → 模型懂的词块多
 🔸 维度高 → 每个词块的语义表达更细腻

两者互不替代，也不能混为一谈。


### target_modules：为什么选这三个名字？

```
target_modules = ["c_attn", "c_fc", "c_proj"]
```

这三个名字，对应的是 **GPT-2 Block 里的几个线性层**，PEFT 会根据名字去“匹配并替换”：

#### 1. `c_attn` — 注意力的 QKV 投影（最关键）

在 GPT-2 里，每一层的自注意力模块大致是：

- 输入 hidden state → 经过 `c_attn` 这个 Linear
- 在这个 Linear 里面，权重被分成三块：Q（Query）、K（Key）、V（Value）
- 所以 `c_attn` 其实是一个 **“QKV 合并投影层”**

👉 在这里加 LoRA 的意义：

- 改变 Q/K/V 的投影方式，相当于改变 **注意力如何“看世界”**
- 对模型行为影响非常大，但参数量相对可控
- 很多实践里，**只对 `c_attn` 加 LoRA 效果就很好，参数极少**

对应你的注释：

> - 只替换 ["c_attn"]: 最少参数，适合显存极小场景

这是最“小成本”的 LoRA 微调方案。

------

#### 2. `c_proj` — 注意力输出投影 + FFN 输出投影

GPT-2 里有两个同名的 `c_proj`：

1. 注意力模块里的 `c_proj`：
   - 把多头注意力的输出，重新映射回隐藏维度
2. 前馈网络（MLP）里的 `c_proj`：
   - 把中间维度（一般更大）映射回隐藏维度

虽然名字都叫 `c_proj`，但：

- 在模型层级结构里，它们有 **不同的路径**（类似 `h.0.attn.c_proj` / `h.0.mlp.c_proj`）
- PEFT 是根据 **完整模块路径** 区分的，不会搞混，只要名字匹配就会各自加 LoRA

👉 在这里加 LoRA 的意义：

- 注意力 `c_proj`：
  - 决定注意力输出如何“融合、回流”到主通道
- MLP `c_proj`：
  - 决定非线性变换之后如何回到原维度

所以：

> - 替换 ["c_attn", "c_proj"]: 平衡选择

这时候 LoRA **覆盖了注意力的输入（QKV）+ 输出（c_proj）+ MLP 输出**，
 既控制“看哪里”（注意力）又控制“最后怎么整合”。

------

#### 3. `c_fc` — FFN（前馈网络）的第一层

在每个 Transformer Block 的 MLP 里：

- `c_fc`: hidden_dim → 4 * hidden_dim（扩大维度）
- 激活函数（比如 GELU）
- `c_proj`: 4 * hidden_dim → hidden_dim（缩回去）

`c_fc` 决定了：

- 模型在“单 token 维度”上的非线性变换前，如何组合特征
- 这个层改变的是 **特征空间的“展开方式”**

在 `c_fc` 上加 LoRA：

- **可以更灵活地调整“特征组合方式”**
- **对下游任务中复杂模式的拟合更强**

因此你这套配置：

> - 替换 ["c_attn", "c_fc", "c_proj"]: 本教程使用，覆盖主要层

实际上是：**注意力 + MLP 的核心线性层都加了一遍 LoRA**
 → 兼顾效果和参数量，是教程里非常典型的一种“全局但不夸张”的方案。

------

#### 4. 为啥“不建议替换 embedding 层和分类头”？

- **Embedding（词向量层）**：
  - 参数量巨大（vocab_size × hidden_dim）
  - **这层更多是“语义坐标系”**，动太多容易破坏预训练学到的通用语义空间
  - 直接 full fine-tune 这层都要小心，更不用说乱插 LoRA
- **分类头（lm_head / output head）**：
  - 通常是一个 Linear：hidden_dim → vocab_size 或类别数
  - 对生成/分类行为影响极大
  - 如果你需要改分类头，通常是 **直接训练这个头本身**，而不是加 LoRA

LoRA 的典型哲学是：

> “尽量在 **中间表征层** 改，而不是在 **输入/输出极端** 大动干戈”

####  综合结论

> `c_attn` 让模型学会**新关注点**，
>  `c_proj` 让模型学会**新输出表达方式**，
>  `c_fc` 让模型获得**新思考能力**。

三者一起覆盖了 **信息获取 → 处理 → 输出** 全链路，

#### 换模型也能准确选 target_modules

不同模型模块名不一样，但 Transformer 结构高度统一。选择目标模块不靠记名字，而靠**识别三类关键线性层**：

| 关键类别                          | 功能                        | LoRA 应放在哪？                                              |
| --------------------------------- | --------------------------- | ------------------------------------------------------------ |
| **Attention QKV 输入投影**        | 模型如何关注信息            | `q_proj / k_proj / v_proj / in_proj / c_attn`                |
| **Attention 输出投影**            | 多头attention合并与输出回流 | `o_proj / out_proj / c_proj / attn.out_proj`                 |
| **FeedForward 前向扩张 + 压缩层** | 抽象推理 & 特征组合         | `fc1 / fc2 / gate_proj / up_proj / down_proj / c_fc / mlp.c_proj` |

#### 📌 常见模型对应 target_modules 参考对照表

| 模型                                             | 推荐 target_modules                                          |
| ------------------------------------------------ | ------------------------------------------------------------ |
| **GPT-2 / GPT-Neo / GPT-J**                      | `["c_attn", "c_proj", "c_fc"]`                               |
| **LLaMA / Qwen / Baichuan / InternLM / Mistral** | `["q_proj", "k_proj", "v_proj", "o_proj"] + ["up_proj","down_proj","gate_proj"]` |
| **BERT / RoBERTa**                               | `["query","key","value","output.dense"] + ["intermediate.dense","output.dense"]` |
| **T5 / Flan-T5**                                 | `["q","k","v","o"] + ["wi","wo"]`（部分模型拆成 `wi_0 + wi_1`） |
| **CLIP / ViT模型**                               | 依然是 QKV + FFN：`["qkv","proj","fc1","fc2"]`               |

📍 你会发现：无论是语言、视觉还是多模态模型，本质选点不变。

------

#### 🧠 通用决策流程（最关键）

未来遇到任何新架构，只要照这个流程走：

```
1. 找 Attention 模块
   - 有 Q/K/V 的 Linear → 必放
   - 有输出投影 → 通常也放

2. 找 FeedForward (MLP) 模块
   - 扩张层 up_proj / wi / fc1
   - 回缩层 down_proj / wo / fc2
   → 推荐至少放其中一段，效果显著提升

3. 扩展选择（进阶）
   - 想更强：MQA/Grouped-Query、RMSNorm可加入
   - 想更轻：只插 Attention，不动 MLP
```

**一句话记住**：

> 无论模型叫什么，LoRA只加在 *Attention + MLP的线性层*，
>  换名字只是换皮肤，作用永远一致。

# 6️⃣ 训练和评估函数

## 🎯 训练流程概述

标准的深度学习训练流程包括：
1. **训练循环**：前向传播 → 计算损失 → 反向传播 → 更新参数
2. **评估循环**：在验证集上计算性能指标
3. **学习率调度**：动态调整学习率
4. **早停机制**：防止过拟合
> 早停（Early Stopping）就是在训练还没过度拟合前提前停止训练，避免模型越学越“背答案”、性能变差。 **当验证集表现不再提升时，就提前让模型停止训练。**


## 📊 PEFT 模型的训练特点

使用 PEFT 后，训练代码与普通 PyTorch 模型**完全相同**！
- ✅ 自动只优化 LoRA 参数
- ✅ 自动冻结原始参数
- ✅ 无需手动筛选参数


In [ ]:
# 📊 评估函数定义
# 功能：计算模型在数据集上的准确率和损失

def evaluate(model, data_loader):
    """
    评估模型性能

    🎯 目的：在验证集或测试集上计算准确率和平均损失

    参数：
        model: 待评估的模型
        data_loader: DataLoader 对象（验证集或测试集）

    返回：
        acc: 准确率 (0-1 之间)
        avg_loss: 平均损失值

    工作流程：
        1. 切换到评估模式 (model.eval())
        2. 禁用梯度计算 (torch.no_grad())
        3. 遍历所有批次计算损失和准确率
        4. 返回平均值
    """
    # 🔧 切换到评估模式
    # 作用：
    # 1. 禁用 Dropout（所有神经元都参与计算）
    # 2. 禁用 Batch Normalization 的更新（使用训练时的统计量）
    # 3. 某些层的行为会改变
    model.eval()

    # 📊 初始化统计变量
    total_loss = 0.0    # 累积损失
    correct = 0         # 正确预测数
    total = 0           # 总样本数

    # 🚫 禁用梯度计算
    # 作用：节省显存，加快计算速度
    # 评估时不需要梯度，因为不进行参数更新
    with torch.no_grad():
        # 🔄 遍历数据批次
        for batch in data_loader:
            # 🎮 将数据移动到设备（GPU/CPU）
            # 字典推导式：将 batch 中的每个张量都移动到目标设备
            batch = {k: v.to(device) for k, v in batch.items()}

            # 🔮 前向传播
            # model(**batch) 等价于 model(input_ids=..., attention_mask=..., labels=...)
            # Hugging Face 模型返回一个包含 loss 和 logits 的对象
            outputs = model(**batch)

            # 📉 提取损失值
            # 当传入 labels 时，模型会自动计算交叉熵损失
            loss = outputs.loss

            # 🎯 获取预测结果
            # logits: [batch_size, num_labels] 形状的张量
            # argmax(dim=-1): 沿最后一维取最大值的索引，得到预测类别
            preds = outputs.logits.argmax(dim=-1)

            # 📊 累积统计量
            total_loss += loss.item()  # 累加损失（转为 Python 数值）
            correct += (preds == batch["labels"]).sum().item()  # 统计正确预测数
            total += batch["labels"].size(0)  # 累加样本数

    # 📐 计算平均值
    avg_loss = total_loss / max(len(data_loader), 1)  # 平均损失
    acc = correct / total if total > 0 else 0.0        # 准确率

    return acc, avg_loss


# 🚀 训练函数定义
# 功能：执行完整的训练流程

def train(model, train_loader, val_loader, epochs=3, lr=5e-4):
    """
    训练 LoRA 模型

    🎯 目的：执行完整的训练循环，包括优化器、学习率调度和梯度裁剪

    参数：
        model: PEFT 模型（已应用 LoRA）
        train_loader: 训练数据加载器
        val_loader: 验证数据加载器
        epochs: 训练轮数（默认 3）
        lr: 学习率（默认 5e-4，即0.0005）

    训练流程：
        每个 epoch:
            1. 训练阶段：遍历训练集，更新参数
            2. 评估阶段：在验证集上计算性能
            3. 学习率调度：根据验证损失调整学习率

    💡 LoRA 微调的学习率建议：
    - 全量微调：1e-5 ~ 5e-5（小学习率）
    - LoRA 微调：5e-4 ~ 1e-3（可以用较大学习率）
    """
    print("🚀 开始训练...")
    print("=" * 60)

    # ⚙️ 创建优化器
    # AdamW: Adam 优化器的改进版，带权重衰减（Weight Decay）
    # model.parameters() 会自动只返回 requires_grad=True 的参数
    # 因此这里会自动只优化 LoRA 参数！
    optimizer = torch.optim.AdamW(model.parameters(), lr=lr)

    print(f"⚙️  优化器: AdamW (lr={lr})")

    # 📅 创建学习率调度器
    # get_linear_schedule_with_warmup: 带预热的线性学习率调度
    #
    # 🔥 Warmup（预热）的作用：
    # - 训练初期使用较小的学习率，逐渐增加到目标学习率
    # - 防止训练初期梯度过大导致的不稳定
    # - 公式：lr = base_lr * (current_step / warmup_steps)
    #
    # 📉 线性衰减：
    # - Warmup 结束后，学习率线性降低到 0
    # - 有助于训练后期的精细调整
    total_steps = epochs * len(train_loader)  # 总训练步数
    warmup_steps = max(10, int(0.1 * total_steps))  # Warmup 步数（10% 的总步数）

    scheduler = get_linear_schedule_with_warmup(
        optimizer,
        num_warmup_steps=warmup_steps,
        num_training_steps=total_steps,
    )

    print(f"📅 学习率调度器: Linear with Warmup")
    print(f"   - 总步数: {total_steps}")
    print(f"   - Warmup 步数: {warmup_steps}")

    # 🔄 训练循环
    for epoch in range(epochs):
        # 📚 训练阶段
        model.train()  # 切换到训练模式
        running_loss = 0.0

        # 📊 使用 tqdm 显示进度条
        pbar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs}")

        for batch in pbar:
            # 🎮 数据移动到设备
            batch = {k: v.to(device) for k, v in batch.items()}

            # 🧹 清空梯度
            # PyTorch 默认累积梯度，必须手动清零
            optimizer.zero_grad()

            # 🔮 前向传播
            outputs = model(**batch)
            loss = outputs.loss

            # 🔄 反向传播
            # 计算所有参数的梯度
            loss.backward()

            # ✂️ 梯度裁剪
            # 限制梯度范数，防止梯度爆炸
            # max_norm=1.0: 如果梯度范数 > 1.0，按比例缩放到 1.0
            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)

            # ⚡ 参数更新
            optimizer.step()      # 使用梯度更新参数
            scheduler.step()      # 更新学习率

            # 📊 累积损失
            running_loss += loss.item()

            # 📈 更新进度条显示
            pbar.set_postfix({'loss': f'{loss.item():.4f}'})

        # 📊 计算本轮平均损失
        avg_train_loss = running_loss / len(train_loader)

        # 📈 验证阶段
        val_acc, val_loss = evaluate(model, val_loader)

        # 📢 打印结果
        print(f"Epoch {epoch+1}/{epochs}:")
        print(f"   - 训练损失: {avg_train_loss:.4f}")
        print(f"   - 验证损失: {val_loss:.4f}")
        print(f"   - 验证准确率: {val_acc*100:.1f}%")
        print(f"   - 当前学习率: {scheduler.get_last_lr()[0]:.2e}")

    print("\n✅ 训练完成！")


# 🎓 初学者提示：
# 以上训练函数集成了多个训练技巧：
# 1. AdamW 优化器：比 SGD 收敛更快，比 Adam 泛化更好
# 2. 学习率预热：训练初期稳定性更好
# 3. 梯度裁剪：防止梯度爆炸
# 4. 进度条显示：实时监控训练状态
#
# 这些都是工业界的标准做法，建议在自己的项目中使用。

print("✅ 训练和评估函数定义完成")


✅ 训练和评估函数定义完成


In [ ]:
# 🚀 执行训练
# 功能：调用训练函数开始微调

# ⚙️ 训练参数设置
EPOCHS = 3          # 训练轮数
LEARNING_RATE = 5e-4  # 学习率

print("⚙️  训练参数:")
print(f"   - 训练轮数: {EPOCHS}")
print(f"   - 学习率: {LEARNING_RATE}")
print(f"   - 批次大小: 8")
print(f"   - 优化器: AdamW")
print(f"   - 学习率调度: Linear with Warmup")
print("")

# 🚀 开始训练
train(model, train_loader, val_loader, epochs=EPOCHS, lr=LEARNING_RATE)

# 📊 在测试集上评估
print("\n📊 最终测试...")
print("=" * 60)
test_acc, test_loss = evaluate(model, test_loader)

print(f"✅ 测试结果:")
print(f"   - 测试损失: {test_loss:.4f}")
print(f"   - 测试准确率: {test_acc*100:.2f}%")

# 🎯 性能分析
if test_acc >= 0.95:
    print("\n🎉 优秀！模型达到了 95%+ 的准确率！")
elif test_acc >= 0.90:
    print("\n👍 很好！模型达到了 90%+ 的准确率！")
elif test_acc >= 0.85:
    print("\n✅ 不错！模型达到了 85%+ 的准确率！")
else:
    print("\n💡 提示：准确率偏低，建议调整超参数或增加训练轮数")


⚙️  训练参数:
   - 训练轮数: 3
   - 学习率: 0.0005
   - 批次大小: 8
   - 优化器: AdamW
   - 学习率调度: Linear with Warmup

🚀 开始训练...
⚙️  优化器: AdamW (lr=0.0005)
📅 学习率调度器: Linear with Warmup
   - 总步数: 393
   - Warmup 步数: 39


Epoch 1/3: 100%|██████████| 131/131 [00:21<00:00,  6.10it/s, loss=0.0001]


Epoch 1/3:
   - 训练损失: 0.7324
   - 验证损失: 0.0785
   - 验证准确率: 98.0%
   - 当前学习率: 3.70e-04


Epoch 2/3: 100%|██████████| 131/131 [00:17<00:00,  7.37it/s, loss=0.0000]


Epoch 2/3:
   - 训练损失: 0.1346
   - 验证损失: 0.0472
   - 验证准确率: 98.0%
   - 当前学习率: 1.85e-04


Epoch 3/3: 100%|██████████| 131/131 [00:16<00:00,  8.06it/s, loss=0.0001]


Epoch 3/3:
   - 训练损失: 0.0675
   - 验证损失: 0.0406
   - 验证准确率: 98.0%
   - 当前学习率: 0.00e+00

✅ 训练完成！

📊 最终测试...
✅ 测试结果:
   - 测试损失: 0.0422
   - 测试准确率: 99.00%

🎉 优秀！模型达到了 95%+ 的准确率！


# 7️⃣ 实际应用：文本分类推理

## 🎯 使用训练好的模型进行预测

现在让我们使用训练好的模型对新文本进行分类，看看效果如何。


In [ ]:
# 🔮 推理演示
# 功能：使用训练好的模型对新文本进行分类预测

# 📝 准备测试样本
# 包含明显的正常短信和垃圾短信
sample_texts = [
    "Hey, want to grab lunch together?",                           # 正常短信
    "URGENT! You've won $5000! Click now to claim your prize!",   # 垃圾短信
    "The meeting has been moved to 2pm",                          # 正常短信
    "Free iPhone! Limited time offer! Call immediately!",         # 垃圾短信
    "Thanks for your help yesterday",                              # 正常短信
    "Congratulations! You are selected as winner. Text WIN now!",  # 垃圾短信
]

print("🔮 推理演示")
print("=" * 60)
print(f"测试 {len(sample_texts)} 条样本...\n")

# 🔧 切换到评估模式
model.eval()

# 🔄 遍历每个测试样本
for i, text in enumerate(sample_texts, 1):
    # 🔤 使用 tokenizer 编码文本
    # 参数说明：
    # - return_tensors="pt": 返回 PyTorch 张量
    # - truncation=True: 如果超长则截断
    # - padding="max_length": 填充到最大长度
    # - max_length: 与训练时保持一致
    encoded = tokenizer(
        text,
        return_tensors="pt",
        truncation=True,
        padding="max_length",
        max_length=max_length,
    )

    # 🎮 将数据移动到设备
    encoded = {k: v.to(device) for k, v in encoded.items()}

    # 🚫 禁用梯度计算（推理时不需要）
    with torch.no_grad():
        # 🔮 前向传播获取 logits
        logits = model(**encoded).logits

        # 📊 计算概率分布
        # softmax 将 logits 转换为概率（和为 1）
        probs = torch.softmax(logits, dim=-1)[0]

    # 🎯 获取预测结果
    pred = torch.argmax(probs).item()  # 预测类别（0 或 1）
    label = "spam" if pred == 1 else "ham"  # 转换为文本标签

    # 📊 提取概率值
    ham_prob = probs[0].item()   # 正常短信概率
    spam_prob = probs[1].item()  # 垃圾短信概率

    # 📢 打印结果
    print(f"【样本 {i}】")
    print(f"文本: {text[:60]}{'...' if len(text) > 60 else ''}")
    print(f"预测: {label.upper()}")
    print(f"置信度: ham={ham_prob:.3f}, spam={spam_prob:.3f}")

    # 🎨 添加表情符号增强可读性
    if pred == 1:
        print("🚨 分类结果：垃圾短信")
    else:
        print("✅ 分类结果：正常短信")

    print()

print("✅ 推理演示完成！")


🔮 推理演示
测试 6 条样本...

【样本 1】
文本: Hey, want to grab lunch together?
预测: HAM
置信度: ham=1.000, spam=0.000
✅ 分类结果：正常短信

【样本 2】
文本: URGENT! You've won $5000! Click now to claim your prize!
预测: SPAM
置信度: ham=0.000, spam=1.000
🚨 分类结果：垃圾短信

【样本 3】
文本: The meeting has been moved to 2pm
预测: SPAM
置信度: ham=0.228, spam=0.772
🚨 分类结果：垃圾短信

【样本 4】
文本: Free iPhone! Limited time offer! Call immediately!
预测: SPAM
置信度: ham=0.000, spam=1.000
🚨 分类结果：垃圾短信

【样本 5】
文本: Thanks for your help yesterday
预测: HAM
置信度: ham=0.982, spam=0.018
✅ 分类结果：正常短信

【样本 6】
文本: Congratulations! You are selected as winner. Text WIN now!
预测: SPAM
置信度: ham=0.000, spam=1.000
🚨 分类结果：垃圾短信

✅ 推理演示完成！


# 8️⃣9️⃣ 模型保存和加载

## 💾 PEFT 模型的保存策略

PEFT 库提供了非常方便的模型保存功能：

### 🎯 两种保存方式

1. **只保存 LoRA 参数**（推荐）
   - 文件很小（几 MB）
   - 需要配合原始模型使用
   - 适合模型分发和多任务场景

2. **保存合并后的完整模型**
   - 将 LoRA 参数合并到原始权重
   - 文件较大
   - 可以作为独立模型使用


In [ ]:
# 💾 方式 1：只保存 LoRA 参数（推荐）
# 功能：保存轻量级的 LoRA 适配器文件

print("💾 保存 LoRA 参数...")

# 📁 指定保存路径
lora_save_path = "./lora_sms_spam_classifier"

# 💾 保存 LoRA 适配器
# save_pretrained() 会保存：
# 1. adapter_config.json: LoRA 配置
# 2. adapter_model.safetensors: LoRA 权重（推荐格式）
# 或 adapter_model.bin: LoRA 权重（传统格式）
model.save_pretrained(lora_save_path)

print(f"✅ LoRA 参数已保存到: {lora_save_path}")

# 📊 检查保存的文件
import os
saved_files = os.listdir(lora_save_path)
print(f"📂 保存的文件: {saved_files}")

# 📏 检查文件大小
total_size = sum(os.path.getsize(os.path.join(lora_save_path, f)) for f in saved_files)
print(f"📊 总大小: {total_size / (1024*1024):.2f} MB")

# 💡 加载方式示例（代码仅供参考，不实际执行）
print("\n💡 加载 LoRA 模型的方法:")
print("```python")
print("from peft import PeftModel")
print("from transformers import AutoModelForSequenceClassification")
print("")
print("# 1. 加载基础模型")
print('base_model = AutoModelForSequenceClassification.from_pretrained("gpt2", num_labels=2)')
print("")
print("# 2. 加载 LoRA 适配器")
print('model = PeftModel.from_pretrained(base_model, "./lora_sms_spam_classifier")')
print("")
print("# 3. 使用模型进行推理")
print("model.eval()")
print("# ... 推理代码 ...")
print("```")

print("\n" + "="*60)

# 🔗 方式 2：合并并保存完整模型（可选）
print("\n🔗 方式 2：合并 LoRA 参数到基础模型")

# 📝 合并说明
print("💡 合并操作会将 LoRA 参数加到原始权重上：")
print("   W_new = W_original + α/r * (A @ B)")
print("")
print("⚠️  注意：合并后将失去切换 LoRA 适配器的能力")
print("")

# 🔧 合并模型（示例代码）
print("🔧 合并代码示例（不执行，仅展示）:")
print("```python")
print("# 合并 LoRA 参数到基础模型")
print("merged_model = model.merge_and_unload()")
print("")
print("# 保存完整模型")
print('merged_model.save_pretrained("./merged_sms_spam_classifier")')
print('tokenizer.save_pretrained("./merged_sms_spam_classifier")')
print("")
print("# 加载完整模型")
print('loaded_model = AutoModelForSequenceClassification.from_pretrained(')
print('    "./merged_sms_spam_classifier"')
print(')')
print("```")

print("\n✅ 模型保存完成！")

# 📊 保存方式对比
print("\n📊 两种保存方式对比:")
print("┌─────────────────┬──────────────┬──────────────┐")
print("│      特性       │  LoRA 参数   │  完整模型    │")
print("├─────────────────┼──────────────┼──────────────┤")
print("│  文件大小       │  几 MB       │  数百 MB     │")
print("│  加载速度       │  快          │  较慢        │")
print("│  多任务切换     │  支持        │  不支持      │")
print("│  独立使用       │  需基础模型  │  可独立使用  │")
print("│  推荐场景       │  生产环境    │  演示/部署   │")
print("└─────────────────┴──────────────┴──────────────┘")


💾 保存 LoRA 参数...
✅ LoRA 参数已保存到: ./lora_sms_spam_classifier
📂 保存的文件: ['adapter_config.json', 'README.md', 'adapter_model.safetensors']
📊 总大小: 4.52 MB

💡 加载 LoRA 模型的方法:
```python
from peft import PeftModel
from transformers import AutoModelForSequenceClassification

# 1. 加载基础模型
base_model = AutoModelForSequenceClassification.from_pretrained("gpt2", num_labels=2)

# 2. 加载 LoRA 适配器
model = PeftModel.from_pretrained(base_model, "./lora_sms_spam_classifier")

# 3. 使用模型进行推理
model.eval()
# ... 推理代码 ...
```


🔗 方式 2：合并 LoRA 参数到基础模型
💡 合并操作会将 LoRA 参数加到原始权重上：
   W_new = W_original + α/r * (A @ B)

⚠️  注意：合并后将失去切换 LoRA 适配器的能力

🔧 合并代码示例（不执行，仅展示）:
```python
# 合并 LoRA 参数到基础模型
merged_model = model.merge_and_unload()

# 保存完整模型
merged_model.save_pretrained("./merged_sms_spam_classifier")
tokenizer.save_pretrained("./merged_sms_spam_classifier")

# 加载完整模型
loaded_model = AutoModelForSequenceClassification.from_pretrained(
    "./merged_sms_spam_classifier"
)
```

✅ 模型保存完成！

📊 两种保存方式对比:
┌─────────────────┬──

# 🎓 教程总结

## ✨ 核心要点

### PEFT 库使用三步曲
1. 创建 LoRAConfig
2. 加载基础模型  
3. 调用 get_peft_model()

### 关键优势
- 代码简洁（~10行 vs ~200行）
- 工业级实现
- 官方维护更新
- 与 Transformers 无缝集成

## 🎯 超参数建议

- **rank**: 8-16（推荐起点）
- **alpha**: 2 × rank
- **dropout**: 0.05-0.1
- **learning_rate**: 5e-4 到 1e-3

## 💡 最佳实践

1. 使用学习率预热（Warmup）
2. 添加梯度裁剪
3. 监控验证集性能
4. 保存轻量级 LoRA 参数

## 🚀 进阶方向

- QLoRA：结合量化
- 多任务 LoRA
- 大模型微调（LLaMA、Mistral）

🎉 恭喜完成学习！
